rust/main/agents/relayer/src/msg/pending_message.rs

```diff
@@ -516,44 +517,37 @@ impl PendingOperation for PendingMessage {
 
 impl PendingMessage {
     /// Constructor that tries reading the retry count from the HyperlaneDB in order to recompute the `next_attempt_after`.
+    /// If the message has been retried more than `max_retries`, it will return `None`.
     /// In case of failure, behaves like `Self::new(...)`.
-    pub fn from_persisted_retries(
+    pub fn maybe_from_persisted_retries(
         message: HyperlaneMessage,
         ctx: Arc<MessageContext>,
         app_context: Option<String>,
-    ) -> Self {
-        // Attempt to fetch status about message from database
-        let message_status = match ctx.origin_db.retrieve_status_by_message_id(&message.id()) {
-            Ok(Some(status)) => {
-                // This event is used for E2E tests to ensure message statuses
-                // are being properly loaded from the db
-                tracing::event!(
-                    if cfg!(feature = "test-utils") {
-                        Level::DEBUG
-                    } else {
-                        Level::TRACE
-                    },
-                    ?status,
-                    id=?message.id(),
-                    RETRIEVED_MESSAGE_LOG,
-                );
-                status
-            }
-            _ => {
-                tracing::event!(
-                    if cfg!(feature = "test-utils") {
-                        Level::DEBUG
-                    } else {
-                        Level::TRACE
-                    },
-                    "Message status not found in db"
-                );
-                PendingOperationStatus::FirstPrepareAttempt
-            }
-        };
+        retries_before_skipping: Option<u32>,
+    ) -> Option<Self> {
+        let num_retries = Self::maybe_get_num_retries(
+            ctx.origin_db.clone(),
+            message.clone(),
+            retries_before_skipping,
+        )?;
+
+        let message_status = Self::get_message_status(ctx.origin_db.clone(), message.clone());
```

Comment:
We can avoid the clone here if `get_message_status` does not take ownership of its `message` parameter.